src/decode/stream.rs

```diff
@@ -0,0 +1,476 @@
+    /// receive decompressed bytes.
+    pub fn new_with_options(options: Options, output: W) -> Self {
+        Self {
+            tmp: [0; MAX_TMP_LEN],
+            tmp_len: 0,
+            state: Some(State::Init(output)),
+            options,
+        }
+    }
+
+    /// Get a reference to the output sink
+    pub fn get_ref(&self) -> Option<&W> {
+        match &self.state {
+            Some(State::Init(output)) => Some(&output),
+            Some(State::Run(state)) => Some(state.decoder.output.get_ref()),
+            None => None,
+        }
+    }
+
+    /// Get a mutable reference to the output sink
+    pub fn get_mut(&mut self) -> Option<&mut W> {
+        match &mut self.state {
+            Some(State::Init(output)) => Some(output),
+            Some(State::Run(state)) => Some(state.decoder.output.get_mut()),
+            None => None,
+        }
+    }
+
+    /// Consumes the stream and returns the output sink. This also makes sure
+    /// we have properly reached the end of the stream.
+    pub fn finish(mut self) -> Result<W, std::io::Error> {
+        if let Some(state) = self.state.take() {
+            match state {
+                State::Init(output) => {
+                    if self.tmp_len > 0 {
+                        Err(std::io::Error::new(
+                            std::io::ErrorKind::Other,
+                            "failed to read header",
+                        ))
+                    } else {
+                        Ok(output)
+                    }
+                }
+                State::Run(mut state) => {
+                    if !self.options.allow_incomplete {
+                        // Process one last time with empty input to force end of
+                        // stream checks
+                        let mut stream = Cursor::new(&self.tmp[0..self.tmp_len]);
+                        let mut range_decoder =
+                            RangeDecoder::from_parts(&mut stream, state.range, state.code);
+                        state
+                            .decoder
+                            .process(&mut range_decoder)
+                            .map_err(|e| -> std::io::Error { e.into() })?;
+                    }
+                    state.decoder.output.finish()
+                }
+            }
+        } else {
+            // this will occur if a call to `write()` fails
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "can't finish stream because of previous write error",
+            ))
+        }
+    }
+
+    /// Attempts to read the header and transition into a running state.
+    ///
+    /// This function will consume the state, returning the next state on both
+    /// error and success.
+    fn read_header<R: BufRead>(
+        output: W,
+        mut input: &mut R,
+        options: &Options,
+    ) -> Result<State<W>, (Option<State<W>>, std::io::Error)> {
```

Comment:
`(Option<State<W>>, std::io::Error)` looks weird for an error type. Can you extend `LZMAError` for this use case?

Reply:
Unfortunately that would require adding a type parameter to `LZMAError`, i.e. `LZMAError<W>`. I think a less intrusive way would be to create an error type inside of `stream.rs` and returning that:
```
Result<State<W>, StreamStateError<W>>
```